File: ui/views/combat_page.py
# -*- coding: utf-8 -*-
import os
from PySide6.QtWidgets import (QWidget, QVBoxLayout, QHBoxLayout, QLabel, QGridLayout, QTextEdit, QFrame)
from PySide6.QtCore import Qt
from PySide6.QtGui import QFont, QPixmap
from ..widgets import Card, StyledButton, StyledProgressBar
from ..theme import Theme

class CombatPage(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.battle_log = ["A wild enemy appears!", "Prepare for battle!"]
        self.init_ui()
        self.setup_accessibility()

    def setup_accessibility(self):
        """Set up accessibility features for the combat page"""
        self.setAccessibleName("Combat Interface")
        self.setAccessibleDescription("Battle arena where you fight enemies")

        # Combat buttons accessibility
        self.attack_btn.setAccessibleName("Attack button")
        self.attack_btn.setAccessibleDescription("Attack the enemy (A)")

        self.use_skill_btn.setAccessibleName("Use skill button")
        self.use_skill_btn.setAccessibleDescription("Use a special skill (S)")

        self.use_item_btn.setAccessibleName("Use item button")
        self.use_item_btn.setAccessibleDescription("Use a consumable item (I)")

        self.run_btn.setAccessibleName("Run away button")
        self.run_btn.setAccessibleDescription("Attempt to escape from battle (R)")

        # Health bars accessibility
        self.hero_combat_health_bar.setAccessibleName("Hero health bar")
        self.hero_combat_health_bar.setAccessibleDescription("Current hero health status")

        self.enemy_combat_health_bar.setAccessibleName("Enemy health bar")
        self.enemy_combat_health_bar.setAccessibleDescription("Current enemy health status")

        # Enable keyboard navigation
        self.setFocusPolicy(Qt.FocusPolicy.StrongFocus)

    def init_ui(self):
        main_layout = QHBoxLayout(self)
        main_layout.setContentsMargins(Theme.MARGIN_LG, Theme.MARGIN_LG, Theme.MARGIN_LG, Theme.MARGIN_LG)
        main_layout.setSpacing(Theme.SPACING_LG)

        # Left Column: Battle Arena
        arena_card = self._create_arena_card()
        main_layout.addWidget(arena_card, 2)

        # Right Column: Battle Log
        log_card = self._create_log_card()
        main_layout.addWidget(log_card, 1)

    def _create_arena_card(self):
        card = Card()
        layout = QVBoxLayout()

        # Title with improved styling
        title = QLabel("⚔️ Battle Arena")
        title_font = QFont()
        title_font.setPointSize(18)
        title_font.setBold(True)
        title.setFont(title_font)
        title.setStyleSheet(f"color: {Theme.PRIMARY.name()}; margin-bottom: {Theme.SPACING_SM}px;")
        title.setAccessibleName("Battle Arena Title")
        title.setAccessibleDescription("Combat zone header")
        layout.addWidget(title)

        # Sprites container with better visual hierarchy
        sprites_container = QFrame()
        sprites_container.setFrameShape(QFrame.Shape.Box)
        sprites_container.setStyleSheet(f"""
            QFrame {{
                border: 2px solid {Theme.BORDER.name()};
                border-radius: {Theme.BORDER_RADIUS_MD}px;
                background-color: {Theme.CARD.name()};
                padding: {Theme.SPACING_MD}px;
            }}
        """)

        sprites_layout = QHBoxLayout(sprites_container)
        sprites_layout.setSpacing(Theme.SPACING_MD)

        # Hero sprite with actual character images
        self.hero_sprite_label = QLabel()
        self.hero_sprite_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        self.hero_sprite_label.setMinimumSize(80, 80)
        self.hero_sprite_label.setStyleSheet(f"""
            QLabel {{
                border-radius: {Theme.BORDER_RADIUS_MD}px;
                background-color: {Theme.SECONDARY.name()};
            }}
        """)
        self.load_hero_sprite()  # Load hero sprite from assets
        self.hero_sprite_label.setToolTip("Your character's sprite during battle")
        self.hero_sprite_label.setAccessibleName("Hero sprite")
        self.hero_sprite_label.setAccessibleDescription("Representation of your character in battle")

        # VS indicator
        vs_label = QLabel("⚔️")
        vs_font = QFont()
        vs_font.setPointSize(24)
        vs_label.setFont(vs_font)
        vs_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        vs_label.setAccessibleName("Versus indicator")

        # Enemy sprite with better formatting
        self.enemy_sprite_label = QLabel()
        self.enemy_sprite_label.setAlignment(Qt.AlignmentFlag.AlignCenter)
        self.enemy_sprite_label.setMinimumSize(80, 80)
        self.enemy_sprite_label.setStyleSheet(f"""
            QLabel {{
                border-radius: {Theme.BORDER_RADIUS_MD}px;
                background-color: {Theme.DESTRUCTIVE.name()};
            }}
        """)
        self.enemy_sprite_label.setToolTip("The enemy's sprite during battle")
        self.enemy_sprite_label.setAccessibleName("Enemy sprite")
        self.enemy_sprite_label.setAccessibleDescription("Representation of the enemy in battle")

        sprites_layout.addWidget(self.hero_sprite_label)
        sprites_layout.addWidget(vs_label, alignment=Qt.AlignmentFlag.AlignCenter)
        sprites_layout.addWidget(self.enemy_sprite_label)

        layout.addWidget(sprites_container)

        # Health Bars with better labels and styling
        health_bars_layout = QVBoxLayout()
        health_bars_layout.setSpacing(Theme.SPACING_SM)

        # Hero health section
        hero_health_layout = QVBoxLayout()
        hero_health_label = QLabel("❤️ Hero Health")
        hero_health_label.setStyleSheet(f"color: {Theme.FOREGROUND.name()}; font-weight: bold;")
        hero_health_label.setAccessibleName("Hero health section")
        hero_health_label.setAccessibleDescription("Hero's current health status")

        self.hero_combat_health_bar = StyledProgressBar(variant="health")
        self.hero_combat_health_bar.setAccessibleName("Hero health progress bar")
        self.hero_combat_health_bar.setAccessibleDescription("Visual indicator of hero's remaining health")

        hero_health_layout.addWidget(hero_health_label)
        hero_health_layout.addWidget(self.hero_combat_health_bar)
        health_bars_layout.addLayout(hero_health_layout)

        # Enemy health section
        enemy_health_layout = QVBoxLayout()
        enemy_health_label = QLabel("💀 Enemy Health")
        enemy_health_label.setStyleSheet(f"color: {Theme.FOREGROUND.name()}; font-weight: bold;")
        enemy_health_label.setAccessibleName("Enemy health section")
        enemy_health_label.setAccessibleDescription("Enemy's current health status")

        self.enemy_combat_health_bar = StyledProgressBar(variant="health")
        self.enemy_combat_health_bar.setAccessibleName("Enemy health progress bar")
        self.enemy_combat_health_bar.setAccessibleDescription("Visual indicator of enemy's remaining health")

        enemy_health_layout.addWidget(enemy_health_label)
        enemy_health_layout.addWidget(self.enemy_combat_health_bar)
        health_bars_layout.addLayout(enemy_health_layout)

        layout.addLayout(health_bars_layout)

        # Action Buttons with improved tooltips
        actions_layout = QGridLayout()
        actions_layout.setSpacing(Theme.SPACING_SM)

        self.attack_btn = StyledButton("⚔️ Attack", variant="primary")
        self.attack_btn.setToolTip("Attack the enemy with your weapon (A)")

        self.use_skill_btn = StyledButton("✨ Use Skill", variant="secondary")
        self.use_skill_btn.setToolTip("Use a special skill or spell (S)")

        self.use_item_btn = StyledButton("🧪 Use Item", variant="secondary")
        self.use_item_btn.setToolTip("Use a consumable item (I)")

        self.view_stats_btn = StyledButton("📖 View Stats", variant="secondary")
        self.view_stats_btn.setToolTip("View detailed enemy statistics and tactics (V)")

        # Add accessibility information for the view stats button
        self.view_stats_btn.setAccessibleName("View enemy stats button")
        self.view_stats_btn.setAccessibleDescription("View detailed enemy statistics and tactics (V)")

        self.run_btn = StyledButton("🏃 Run", variant="destructive")
        self.run_btn.setToolTip("Attempt to flee from battle (R)")

        actions_layout.addWidget(self.attack_btn, 0, 0)
        actions_layout.addWidget(self.use_skill_btn, 0, 1)
        actions_layout.addWidget(self.use_item_btn, 1, 0)
        actions_layout.addWidget(self.view_stats_btn, 1, 1)
        actions_layout.addWidget(self.run_btn, 2, 0, 1, 2)

        layout.addLayout(actions_layout)

        # Set layout on card and return
        card.setLayout(layout)
        return card

    def load_hero_sprite(self):
        """Load and display hero sprite image with proper Qt PySide scaling"""
        try:
            # Try to load Hero.png from assets
            hero_sprite_path = "assets/Hero.png"
            if os.path.exists(hero_sprite_path):
                self.hero_pixmap = QPixmap(hero_sprite_path)
                if not self.hero_pixmap.isNull():
                    # Don't scale immediately - wait for layout and resize events
                    # This follows Qt best practices for image scaling
                    self.rescale_hero_sprite()
                else:
                    print("Warning: Failed to load hero sprite pixmap")
                    self.hero_sprite_label.setText("🛡️ HERO")
            else:
                print(f"Warning: Hero sprite not found at {hero_sprite_path}")
                self.hero_sprite_label.setText("🛡️ HERO")
        except Exception as e:
            print(f"Warning: Error loading hero sprite: {e}")
            self.hero_sprite_label.setText("🛡️ HERO")

    def rescale_hero_sprite(self):
        """Properly scale hero sprite according to Qt PySide best practices"""
        if not hasattr(self, 'hero_pixmap') or self.hero_pixmap.isNull():
            return

        # Get the actual container dimensions for proper scaling
        # Use the parent container's size for more reliable calculations
        parent = self.hero_sprite_label.parentWidget()
        container_size = parent.size() if parent else self.hero_sprite_label.size()

        # Calculate target size based on container with proper padding
        # Use a reasonable size that's responsive but not too large
        available_width = max(60, min(140, container_size.width() - Theme.SPACING_MD * 2))
        available_height = max(60, min(140, container_size.height() - Theme.SPACING_MD * 2))

        target_size = min(available_width, available_height)
        target_size = max(target_size, 48)  # Minimum 48px for visibility

        try:
            # Scale with aspect ratio preserved - Qt PySide best practice
            scaled_pixmap = self.hero_pixmap.scaled(
                target_size, target_size,
                Qt.AspectRatioMode.KeepAspectRatio,
                Qt.TransformationMode.SmoothTransformation
            )

            self.hero_sprite_label.setPixmap(scaled_pixmap)
            self.hero_sprite_label.setScaledContents(False)
            self.hero_sprite_label.setAlignment(Qt.AlignmentFlag.AlignCenter)

        except Exception as e:
            print(f"Warning: Error scaling hero sprite: {e}")
            self.hero_sprite_label.setText("🛡️ HERO")

    def resizeEvent(self, event):
        """Handle resize events to rescale sprites properly"""
        super().resizeEvent(event)
        # Rescale hero sprite when container resizes - Qt best practice
        if hasattr(self, 'hero_sprite_label') and hasattr(self, 'hero_pixmap'):
            self.rescale_hero_sprite()

    def load_enemy_sprite(self, enemy):
        """Load and display enemy sprite based on enemy type"""
        try:
            # Map common enemy names to asset files
            enemy_sprite_map = {
                'Goblin': 'goblin.png',
                'Orc': 'Orc.png',
                'Golem': 'golem.png',
                'Mage': 'mage.png',
                'Rogue': 'rogue.png'
            }

            enemy_name = getattr(enemy, 'name', '').split(' ')[0]  # Get first word (type)
            sprite_file = enemy_sprite_map.get(enemy_name, 'goblin.png')  # Default to goblin
            enemy_sprite_path = f"assets/{sprite_file}"

            if os.path.exists(enemy_sprite_path):
                pixmap = QPixmap(enemy_sprite_path)
                if not pixmap.isNull():
                    # Get the actual size for proper scaling
                    target_size = self.enemy_sprite_label.size()
                    scaled_pixmap = pixmap.scaled(
                        target_size,
                        Qt.AspectRatioMode.KeepAspectRatio,
                        Qt.TransformationMode.SmoothTransformation
                    )
                    self.enemy_sprite_label.setPixmap(scaled_pixmap)
                    self.enemy_sprite_label.setScaledContents(False)
                else:
                    print(f"Warning: Failed to load enemy sprite pixmap for {enemy_name}")
                    self.enemy_sprite_label.setText(f"👹 {getattr(enemy, 'name', 'ENEMY')}")
            else:
                print(f"Warning: Enemy sprite not found at {enemy_sprite_path}")
                self.enemy_sprite_label.setText(f"👹 {getattr(enemy, 'name', 'ENEMY')}")
        except Exception as e:
            print(f"Warning: Error loading enemy sprite: {e}")
            self.enemy_sprite_label.setText(f"👹 {getattr(enemy, 'name', 'ENEMY')}")

    def _create_log_card(self):
        card = Card()
        layout = QVBoxLayout()

        # Title with styling
        title = QLabel("📜 Battle Log")
        title_font = QFont()
        title_font.setPointSize(18)
        title_font.setBold(True)
        title.setFont(title_font)
        title.setStyleSheet(f"color: {Theme.PRIMARY.name()}; margin-bottom: {Theme.SPACING_SM}px;")
        title.setAccessibleName("Battle Log Title")
        title.setAccessibleDescription("Combat events and messages header")
        layout.addWidget(title)

        # Battle log display with improved styling
        self.battle_log_display = QTextEdit()
        self.battle_log_display.setReadOnly(True)
        self.battle_log_display.setStyleSheet(f"""
            QTextEdit {{
                background-color: {Theme.CARD.name()};
                color: {Theme.FOREGROUND.name()};
                border: 1px solid {Theme.BORDER.name()};
                border-radius: {Theme.BORDER_RADIUS_MD}px;
                font-family: monospace;
                font-size: 12px;
                padding: {Theme.SPACING_SM}px;
            }}
        """)
        self.battle_log_display.setAccessibleName("Battle log")
        self.battle_log_display.setAccessibleDescription("Chronological list of combat events and messages")
        layout.addWidget(self.battle_log_display)

        card.setLayout(layout)
        return card

    def update_combat_ui(self, hero, enemy):
        if not hero or not enemy:
            return

        # Update hero health with animation
        hero_health_percent = int((hero.health / hero.max_health) * 100)
        self.hero_combat_health_bar.setValue(hero_health_percent)
        self.hero_combat_health_bar.setFormat(f"{hero.health} / {hero.max_health} HP")

        # Update enemy health with animation
        enemy_health_percent = int((enemy.health / enemy.max_health) * 100)
        self.enemy_combat_health_bar.setValue(enemy_health_percent)
        self.enemy_combat_health_bar.setFormat(f"{enemy.health} / {enemy.max_health} HP")

        # Load dynamic sprites for enemies
        self.load_enemy_sprite(enemy)

        self.update_battle_log_display()

    def add_battle_log_message(self, message):
        self.battle_log.append(message)
        if len(self.battle_log) > 20:
            self.battle_log = self.battle_log[-20:]
        self.update_battle_log_display()

    def update_battle_log_display(self):
        if hasattr(self, 'battle_log_display'):
            log_html = f"<div style='line-height: 1.6; color: {Theme.FOREGROUND.name()};'>"
            for msg in self.battle_log[-10:]:
                # Color messages based on content
                if any(word in msg.lower() for word in ["critical", "damage", "attack"]):
                    msg_color = Theme.PRIMARY.name()
                elif any(word in msg.lower() for word in ["defeated", "victory"]):
                    msg_color = Theme.ACCENT.name()
                elif any(word in msg.lower() for word in ["healing", "restored", "recovered"]):
                    msg_color = Theme.SECONDARY.name()
                else:
                    msg_color = Theme.MUTED_FOREGROUND.name()

                log_html += f"<div style='border-left: 3px solid {Theme.PRIMARY.name()}; padding-left: 8px; margin: 4px 0; color: {msg_color};'>{msg}</div>"
            log_html += "</div>"
            self.battle_log_display.setHtml(log_html)

            # Auto-scroll to bottom
            scrollbar = self.battle_log_display.verticalScrollBar()
            scrollbar.setValue(scrollbar.maximum())

File: cpp-qt-rpg/src/views/CombatPage.cpp
#include "CombatPage.h"
#include "../theme/Theme.h"
#include "../models/Player.h"
#include "../models/Monster.h"
#include "../components/ParticleSystem.h"
#include <QHBoxLayout>
#include <QVBoxLayout>
#include <QGridLayout>
#include <QLabel>
#include <QProgressBar>
#include <QPushButton>
#include <QTextEdit>
#include <QFrame>
#include <QFont>

CombatPage::CombatPage(QWidget *parent) : QWidget(parent), m_inCombat(false)
{
    setupUi();

    // Initialize particle system after UI is set up
    m_particleSystem = new ParticleSystem(this);
    m_particleSystem->setGeometry(rect());
    m_particleSystem->raise(); // Ensure particles appear on top
}

void CombatPage::setupUi()
{
    QHBoxLayout *mainLayout = new QHBoxLayout(this);
    mainLayout->setContentsMargins(Theme::SPACING_LG, Theme::SPACING_LG, Theme::SPACING_LG, Theme::SPACING_LG);
    mainLayout->setSpacing(Theme::SPACING_LG);

    QWidget *arenaCard = createArenaCard();
    mainLayout->addWidget(arenaCard, 2);

    QWidget *logCard = createLogCard();
    mainLayout->addWidget(logCard, 1);

    // Start in non-combat mode
    setCombatMode(false);
}

QWidget* CombatPage::createArenaCard()
{
    QFrame *card = new QFrame();
    card->setStyleSheet(QString("background-color: %1; border-radius: %2px;").arg(Theme::CARD.name()).arg(Theme::BORDER_RADIUS_MD));
    m_cardLayout = new QVBoxLayout(card);

    QLabel *title = new QLabel("⚔️ Battle Arena");
    QFont titleFont;
    titleFont.setPointSize(18);
    titleFont.setBold(true);
    title->setFont(titleFont);
    title->setStyleSheet(QString("color: %1; margin-bottom: %2px;").arg(Theme::PRIMARY.name()).arg(Theme::SPACING_SM));
    m_cardLayout->addWidget(title);

    QFrame *spritesContainer = new QFrame();
    spritesContainer->setFrameShape(QFrame::Box);
    spritesContainer->setStyleSheet(QString(
        "QFrame { border: 2px solid %1; border-radius: %2px; background-color: %3; padding: %4px; }"
    ).arg(Theme::BORDER.name()).arg(Theme::BORDER_RADIUS_MD).arg(Theme::CARD.name()).arg(Theme::SPACING_MD));
    QHBoxLayout *spritesLayout = new QHBoxLayout(spritesContainer);
    spritesLayout->setSpacing(Theme::SPACING_MD);

    m_heroSpriteLabel = new QLabel("🛡️ HERO");
    m_heroSpriteLabel->setAlignment(Qt::AlignCenter);
    m_heroSpriteLabel->setMinimumSize(100, 100);
    m_enemySpriteLabel = new QLabel("👹 ENEMY");
    m_enemySpriteLabel->setAlignment(Qt::AlignCenter);
    m_enemySpriteLabel->setMinimumSize(100, 100);

    spritesLayout->addWidget(m_heroSpriteLabel);
    spritesLayout->addWidget(new QLabel("⚔️"));
    spritesLayout->addWidget(m_enemySpriteLabel);
    m_cardLayout->addWidget(spritesContainer);

    // Monster name/level display
    m_monsterNameLabel = new QLabel("Unknown Monster");
    m_monsterNameLabel->setAlignment(Qt::AlignCenter);
    m_monsterNameLabel->setStyleSheet(QString("font-size: 16px; font-weight: bold; color: %1;").arg(Theme::FOREGROUND.name()));
    m_cardLayout->addWidget(m_monsterNameLabel);

    // Hero stats
    m_cardLayout->addWidget(new QLabel("❤️ Hero Health"));
    m_heroHealthBar = new QProgressBar();
    m_heroHealthBar->setRange(0, 100);
    m_heroHealthBar->setValue(100);
    m_heroHealthBar->setTextVisible(true);
    m_heroHealthBar->setFormat("100 / 100 HP");
    m_heroHealthBar->setStyleSheet("QProgressBar { border: 2px solid grey; border-radius: 5px; text-align: center; } QProgressBar::chunk { background-color: #dc3545; }");
    m_cardLayout->addWidget(m_heroHealthBar);

    m_cardLayout->addWidget(new QLabel("💙 Hero Mana"));
    m_heroManaBar = new QProgressBar();
    m_heroManaBar->setRange(0, 100);
    m_heroManaBar->setValue(100);
    m_heroManaBar->setTextVisible(true);
    m_heroManaBar->setFormat("100 / 100 MP");
    m_heroManaBar->setStyleSheet("QProgressBar { border: 2px solid grey; border-radius: 5px; text-align: center; } QProgressBar::chunk { background-color: #3b82f6; }");
    m_cardLayout->addWidget(m_heroManaBar);

    // Enemy stats
    m_cardLayout->addWidget(new QLabel("💀 Enemy Health"));
    m_enemyHealthBar = new QProgressBar();
    m_enemyHealthBar->setRange(0, 100);
    m_enemyHealthBar->setValue(100);
    m_enemyHealthBar->setTextVisible(true);
    m_enemyHealthBar->setFormat("100 / 100 HP");
    m_enemyHealthBar->setStyleSheet("QProgressBar { border: 2px solid grey; border-radius: 5px; text-align: center; } QProgressBar::chunk { background-color: #ef4444; }");
    m_cardLayout->addWidget(m_enemyHealthBar);

    // Actions layout will be added in setCombatMode
    m_actionsLayout = nullptr;

    return card;
}

void CombatPage::updateCombatState(Player *player, Monster *monster, const QString &log)
{
    if (player) {
        m_heroHealthBar->setMaximum(player->maxHealth);
        m_heroHealthBar->setValue(player->health);
        m_heroHealthBar->setFormat(QString("%1 / %2 HP").arg(player->health).arg(player->maxHealth));

        m_heroManaBar->setMaximum(player->maxMana);
        m_heroManaBar->setValue(player->mana);
        m_heroManaBar->setFormat(QString("%1 / %2 MP").arg(player->mana).arg(player->maxMana));

        // Try to load character sprite
        QString spritePath = QString(":/assets/%1.png").arg(player->characterClass.toLower());
        QPixmap sprite(spritePath);
        if (!sprite.isNull()) {
            m_heroSpriteLabel->setPixmap(sprite.scaled(100, 100, Qt::KeepAspectRatio, Qt::SmoothTransformation));
        }
    }
    if (monster) {
        m_enemyHealthBar->setMaximum(monster->maxHealth);
        m_enemyHealthBar->setValue(monster->health);
        m_enemyHealthBar->setFormat(QString("%1 / %2 HP").arg(monster->health).arg(monster->maxHealth));

        // Update monster name and level
        m_monsterNameLabel->setText(QString("%1 (Level %2)").arg(monster->name).arg(monster->level));

        // Try to load monster sprite based on enemyType
        QString spritePath = QString(":/assets/%1.png").arg(monster->enemyType);
        QPixmap sprite(spritePath);
        if (!sprite.isNull()) {
            m_enemySpriteLabel->setPixmap(sprite.scaled(100, 100, Qt::KeepAspectRatio, Qt::SmoothTransformation));
        }
    }
    if (!log.isEmpty()) {
        m_battleLog->append(log);
    }
}

void CombatPage::setCombatActive(bool active)
{
    if (m_inCombat && m_actionsLayout) {
        // Find and enable/disable combat buttons in the current layout
        for (int i = 0; i < m_actionsLayout->count(); ++i) {
            QLayoutItem *item = m_actionsLayout->itemAt(i);
            if (item && item->widget()) {
                QPushButton *button = qobject_cast<QPushButton*>(item->widget());
                if (button) {
                    button->setEnabled(active);
                }
            }
        }
    }
}

QWidget* CombatPage::createLogCard()
{
    QFrame *card = new QFrame();
    card->setStyleSheet(QString("background-color: %1; border-radius: %2px;").arg(Theme::CARD.name()).arg(Theme::BORDER_RADIUS_MD));
    QVBoxLayout *layout = new QVBoxLayout(card);

    QLabel *title = new QLabel("📜 Battle Log");
    QFont titleFont;
    titleFont.setPointSize(18);
    titleFont.setBold(true);
    title->setFont(titleFont);
    title->setStyleSheet(QString("color: %1; margin-bottom: %2px;").arg(Theme::PRIMARY.name()).arg(Theme::SPACING_SM));
    layout->addWidget(title);

    m_battleLog = new QTextEdit();
    m_battleLog->setReadOnly(true);
    m_battleLog->setText("A wild enemy appears!\nPrepare for battle!");
    layout->addWidget(m_battleLog);

    return card;
}

QLayout* CombatPage::createCombatActionsLayout()
{
    QGridLayout *actionsLayout = new QGridLayout();
    actionsLayout->setSpacing(Theme::SPACING_SM);

    QPushButton *attackButton = new QPushButton("⚔️ Attack");
    connect(attackButton, &QPushButton::clicked, this, &CombatPage::attackClicked);
    QPushButton *skillButton = new QPushButton("✨ Use Skill");
    connect(skillButton, &QPushButton::clicked, this, &CombatPage::skillClicked);
    QPushButton *itemButton = new QPushButton("🧪 Use Item");
    connect(itemButton, &QPushButton::clicked, this, &CombatPage::itemClicked);
    QPushButton *statsButton = new QPushButton("📖 View Stats");
    connect(statsButton, &QPushButton::clicked, this, &CombatPage::statsClicked);
    QPushButton *runButton = new QPushButton("🏃 Run");
    connect(runButton, &QPushButton::clicked, this, &CombatPage::runClicked);

    actionsLayout->addWidget(attackButton, 0, 0);
    actionsLayout->addWidget(skillButton, 0, 1);
    actionsLayout->addWidget(itemButton, 1, 0);
    actionsLayout->addWidget(statsButton, 1, 1);
    actionsLayout->addWidget(runButton, 2, 0, 1, 2);

    return actionsLayout;
}

QLayout* CombatPage::createNonCombatActionsLayout()
{
    QVBoxLayout *actionsLayout = new QVBoxLayout();
    actionsLayout->setSpacing(Theme::SPACING_MD);

    QLabel *title = new QLabel("Choose your next action");
    QFont titleFont;
    titleFont.setPointSize(16);
    titleFont.setBold(true);
    title->setFont(titleFont);
    title->setStyleSheet(QString("color: %1; margin-bottom: %2px;").arg(Theme::PRIMARY.name()).arg(Theme::SPACING_SM));
    title->setAlignment(Qt::AlignCenter);
    actionsLayout->addWidget(title);

    QLabel *shortcutsHint = new QLabel("Keyboard shortcuts: I (Inventory), S (Shop), L (Save/Load)");
    shortcutsHint->setStyleSheet(QString("color: %1; font-size: 12px;").arg(Theme::MUTED_FOREGROUND.name()));
    shortcutsHint->setAlignment(Qt::AlignCenter);
    actionsLayout->addWidget(shortcutsHint);

    QPushButton *exploreButton = new QPushButton("🏞️ Explore");
    exploreButton->setMinimumHeight(60);
    exploreButton->setStyleSheet(QString(
        "QPushButton { background-color: %1; color: %2; border-radius: 8px; padding: 10px; font-size: 16px; font-weight: bold; }"
        "QPushButton:hover { background-color: %3; }"
    ).arg(Theme::PRIMARY.name()).arg(Theme::PRIMARY_FOREGROUND.name()).arg(Theme::PRIMARY.darker(110).name()));
    connect(exploreButton, &QPushButton::clicked, this, &CombatPage::exploreClicked);

    QPushButton *restButton = new QPushButton("😴 Rest");
    restButton->setMinimumHeight(60);
    restButton->setStyleSheet(QString(
        "QPushButton { background-color: %1; color: %2; border-radius: 8px; padding: 10px; font-size: 16px; font-weight: bold; }"
        "QPushButton:hover { background-color: %3; }"
    ).arg(Theme::SECONDARY.name()).arg(Theme::FOREGROUND.name()).arg(Theme::SECONDARY.darker(110).name()));
    connect(restButton, &QPushButton::clicked, this, &CombatPage::restClicked);

    QPushButton *menuButton = new QPushButton("📋 Menu");
    menuButton->setMinimumHeight(60);
    menuButton->setStyleSheet(QString(
        "QPushButton { background-color: %1; color: %2; border-radius: 8px; padding: 10px; font-size: 16px; font-weight: bold; }"
        "QPushButton:hover { background-color: %3; }"
    ).arg(Theme::ACCENT.name()).arg(Theme::ACCENT_FOREGROUND.name()).arg(Theme::ACCENT.darker(110).name()));
    connect(menuButton, &QPushButton::clicked, this, &CombatPage::menuClicked);

    actionsLayout->addWidget(exploreButton);
    actionsLayout->addWidget(restButton);
    actionsLayout->addWidget(menuButton);
    actionsLayout->addStretch();

    return actionsLayout;
}

void CombatPage::setCombatMode(bool inCombat)
{
    if (m_inCombat == inCombat && m_actionsLayout != nullptr) return;

    m_inCombat = inCombat;

    // Clear all items from the card layout (except the fixed ones)
    // We need to be careful to only remove the dynamic actions layout
    if (m_actionsLayout) {
        // Find and remove the actions layout from the card layout
        for (int i = m_cardLayout->count() - 1; i >= 0; --i) {
            QLayoutItem *item = m_cardLayout->itemAt(i);
            if (item && item->layout() == m_actionsLayout) {
                m_cardLayout->removeItem(item);
                // Don't delete the item yet, we need to clean up the layout first
                break;
            }
        }

        // Clear and delete the old layout
        QLayoutItem *childItem;
        while ((childItem = m_actionsLayout->takeAt(0)) != nullptr) {
            if (childItem->widget()) {
                delete childItem->widget();
            }
            delete childItem;
        }
        delete m_actionsLayout;
        m_actionsLayout = nullptr;
    }

    // Create new actions based on mode
    if (inCombat) {
        m_actionsLayout = createCombatActionsLayout();
        m_heroSpriteLabel->show();
        m_heroManaBar->show();
        m_enemySpriteLabel->show();
        m_monsterNameLabel->show();
        m_enemyHealthBar->show();
        m_battleLog->setText("A wild enemy appears!\nPrepare for battle!");
    } else {
        m_actionsLayout = createNonCombatActionsLayout();
        m_heroSpriteLabel->hide();
        m_heroManaBar->hide();
        m_enemySpriteLabel->hide();
        m_monsterNameLabel->hide();
        m_enemyHealthBar->hide();
        m_battleLog->setText("Welcome to the Realm of Legends!\nWhat would you like to do?");
    }

    // Add the new layout to the card
    m_cardLayout->addLayout(m_actionsLayout);

    // Force the widget to update its layout
    m_cardLayout->update();
    update();
}

    

QLabel* CombatPage::getHeroSpriteLabel()
{
    return m_heroSpriteLabel;
}

    

QLabel* CombatPage::getEnemySpriteLabel()
{
    return m_enemySpriteLabel;
}

ParticleSystem* CombatPage::getParticleSystem()
{
    return m_particleSystem;
}

void CombatPage::addLogEntry(const QString &message, const QString &type)
{
    // Simple implementation - just append to the battle log
    // The type parameter is optional and could be used for color-coding in the future
    if (m_battleLog) {
        m_battleLog->append(message);
    }
}

